### PR TITLE
Decouple input processing from Term

### PR DESCRIPTION
**Pull request by jwilm**
Should make input processing much more easily tested. Going to clean this up a bit in the morning before merging.

cc @xyin96

### CODE REVIEW

**Comment by xyin96** [comment]
I'm curious as to why you decided to move the `mouse` outside of `ActionContext` and into the `Processor`. I think it might make more sense to have `get_mouse()` on the `ActionContext`.

**Comment by jwilm** [comment]
The `Mouse` is essentially just state for the input processor. The `ActionContext` is intended to be a type for dispatching actions to. The `input::Processor` is an ephemeral type anyhow, and the `Mouse` and `ActionContext` are owned by the caller.

What advantage does having the `Mouse` as part of the `ActionContext` have? It seems like any relevant `Mouse` state could just be passed into methods on it as needed.

File: src/input.rs [hunk 1]
```diff
@@ -240,52 +239,43 @@ impl<'a, N: Notify + 'a> Processor<'a, N> {
...
+            self.ctx.write_to_pty(msg);
         }
     }
 
     pub fn sgr_mouse_report(&mut self, button: u8, release: bool) {
-        let (line, column) = (self.ctx.mouse.line, self.ctx.mouse.column);
+        let (line, column) = (self.mouse.line, self.mouse.column);
         let c = if release { 'm' } else { 'M' };
 
         let msg = format!("\x1b[<{};{};{}{}", button, column + 1, line + 1, c);
```
Comment (lukaslueg) [inline src/input.rs]: Maybe we should have some consts defined for ANSI escapes and `concat!(ANSI_CSI, format!(...)` so people can understand what this is supposed to do (including `m` and `M` in this context)

File: src/input.rs [hunk 2]
```diff
@@ -175,13 +173,13 @@ impl Action {
...
     }
 
-    fn paste<'a, N: Notify>(&self, ctx: &mut ActionContext<'a, N>, contents: String) {
-        if ctx.terminal.mode().contains(mode::BRACKETED_PASTE) {
-            ctx.notifier.notify(&b"\x1b[200~"[..]);
-            ctx.notifier.notify(contents.into_bytes());
-            ctx.notifier.notify(&b"\x1b[201~"[..]);
+    fn paste<A: ActionContext>(&self, ctx: &mut A, contents: String) {
+        if ctx.terminal_mode().contains(mode::BRACKETED_PASTE) {
+            ctx.write_to_pty(&b"\x1b[200~"[..]);
```
Comment (lukaslueg) [inline src/input.rs]: Same thing: Escape sequences could be const-ified

**Comment by xyin96** [comment]
I agree that the functionally, having the `Mouse` as part of the `ActionContext` has no real benefit, but from a sort of design perspective, `ActionContext` seems to hold the relevant information of the most recently dispatched action, while `Processor` (`struct`) seems to hold mainly configuration and a pointer to the context at which it references to consumes the action.

**Comment by jwilm** [comment]
@xyin96 

I don't really feel strongly either way about the `Mouse` thing, and it was on `ActionContext` before. `Mouse` is now back in `ActionContext`, and the trait gained a `mouse_mut(&mut self) -> &mut Mouse` method.

After this is merged, you might want to squash your current WIP before rebasing or you'll probably end up with tons of conflicts on each commit.

File: src/input.rs (hunk 1 above)
Comment (jwilm) [inline src/input.rs]: Agree that would be nice. I'll open an issue.

**Comment by jwilm** [comment]
@xyin96 is this going to work for your needs?

Travis as a whole seems broken, but the tests pass locally. We can merge at any time.

**Comment by xyin96** [comment]
Yeah, it looks good. Thanks! 👍

**Pull request merged**